keps/sig-network/2136-np-version-status/README.md

```diff
@@ -0,0 +1,789 @@
+may work by simply translating NetworkPolicy rules into rules for this
+other engine. As in the distributed case, it may not be possible for
+such a NetworkPolicy implementation to know for sure when a policy was
+fully in effect.
+
+### Risks and Mitigations
+
+## Design Details
+
+### API
+
+Add a new field to `networkingv1.NetworkPolicySpec`:
+
+```
+type NetworkPolicySpec struct {
+	...
+
+	// minVersion indicates the Kubernetes release version corresponding to
+	// the NetworkPolicy functionality required by this policy. Currently the
+	// highest allowed value is "1.12", corresponding to the NetworkPolicy API
+	// as defined in Kubernetes 1.12.
+	// +optional
+	MinVersion NetworkPolicyVersion `json:"minVersion,omitempty" protobuf:"bytes,5,name=minVersion"`
+
+	...
+}
+
+type NetworkPolicyVersion string
+
+const (
+	// Kubernetes 1.3 introduced the NetworkPolicy v1beta1 API. Plugins implementing
+	// this version recognize ingress-only policies containing podSelectors and
+	// namespaceSelectors, and TCP and UDP ports.
+	NetworkPolicyVersion1_3 NetworkPolicyVersion = "1.3"
+
+	// Kubernetes 1.8 brought the NetworkPolicy API to v1, adding egress policies
+	// and ipBlocks.
+	NetworkPolicyVersion1_8 NetworkPolicyVersion = "1.8"
+
+	// Kubernetes 1.9 added alpha IPv6 support. Plugins implementing this version are
+	// assumed to recognize IPv6 addresses, even if they don't implement them.
+	NetworkPolicyVersion1_9 NetworkPolicyVersion = "1.9"
+
+	// Kubernetes 1.11 added support for specifying podSelector and namespaceSelector
+	// together in a peer.
+	NetworkPolicyVersion1_11 NetworkPolicyVersion = "1.11"
+
+	// Kubernetes 1.12 added support for SCTP.
+	NetworkPolicyVersion1_12 NetworkPolicyVersion = "1.12"
+)
+
+```
+
+And add a new `Status NetworkPolicyStatus` field to `NetworkPolicy`, defined as:
+
+```
+// NetworkPolicyStatus contains information about the processing of a NetworkPolicy
+type NetworkPolicyStatus struct {
+	// conditions associated with the policy. These may include a "Supported"
+	// condition indicating whether the features used by the policy are supported,
+	// an "Enforcing" condition indicating whether the policy is actively being
+	// enforced on network traffic yet, and a "Problem" condition indicating some
+	// non-fatal problem with the policy.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions []NetworkPolicyStatusCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,3,rep,name=conditions"`
+}
+
+// NetworkPolicyStatusConditionType is a valid value of NetworkPolicyStatusCondition.Type
+type NetworkPolicyStatusConditionType string
+
+const (
+	// NetworkPolicyStatusSupported - The plugin supports this NetworkPolicy.
+	NetworkPolicyStatusSupported NetworkPolicyStatusConditionType = "Supported"
+
+	// NetworkPolicyStatusProblem - The NetworkPolicy has a problem
+	NetworkPolicyStatusProblem NetworkPolicyStatusConditionType = "Problem"
```

Comment:
`Failed`  maybe  ?since this is the np  equivalent to a pod that wont ever come up

Reply:
`Problem` is supposed to be for the case of like https://github.com/kubernetes/kubernetes/pull/94484 where we're required to enforce a particular policy but the spec doesn't make it 100% clear how the policy should be interpreted. So the plugin could say

    conditions:
    - type: Problem
      status: True
      reason: Ambiguous
      message: "Interpreting 192.168.1.5/24 as 192.168.1.0/24 rather than 192.168.1.5/32"
    - type: Supported
      status: True
    - type: Enforcing
      status: True

"the np equivalent to a pod that wont ever come up" is "not Supported and not Enforcing". eg:

    conditions:
    - type: Supported
      status: False
      reason: Unimplemented
      message: "egress policies are not supported by openshift-sdn"
    - type: Enforcing
      status: False
      reason: Unimplemented
      message: "egress policies are not supported by openshift-sdn"

Reply:
yeah, per debate below, makes sense now .  Warning as a potential alternative name